src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -0,0 +1,176 @@
+
+    /**
+     * Determine and estimate the size of a .git cached directory
+     * @param remoteName: Use the repository url to access a cached Jenkins directory, we do not lock it.
+     * @return useCache
+     * @throws IOException
+     * @throws InterruptedException
+     */
+    private boolean decideAndUseCache(String remoteName) throws IOException, InterruptedException {
+        boolean useCache = false;
+        String cacheEntry = getCacheEntry(remoteName);
+        File cacheDir = AbstractGitSCMSource.getCacheDir(cacheEntry);
+        if (cacheDir != null) {
+            Git git = Git.with(TaskListener.NULL, new EnvVars(EnvVars.masterEnvVars)).in(cacheDir).using("git");
+            GitClient client = git.getClient();
+            if (client.hasGitRepo()) {
+                sizeOfRepo = FileUtils.sizeOfDirectory(cacheDir);
+                sizeOfRepo = (sizeOfRepo/1000); // Conversion from Bytes to Kilo Bytes
+                useCache = true;
+            }
+        }
+        return useCache;
+    }
+
+    private void decideAndUseAPI(String remoteName, Item context, String credentialsId) {
+        if (setSizeFromAPI(remoteName, context, credentialsId)) {
+            implementation = determineSwitchOnSize(sizeOfRepo);
+        }
+    }
+
+    private String getCacheEntry(String remote) {
+        return "git-" + Util.getDigestOf(remote);
+    }
+
+    /**
+     * Check if the desired implementation of extension is present and ask for the size of repository if it does
+     * @param repoUrl: The remote name derived from {@link GitSCMSource} object
+     * @return boolean useAPI or not.
+     */
+    private boolean setSizeFromAPI(String repoUrl, Item context, String credentialsId) {
+        List<RepositorySizeAPI> acceptedRepository = Objects.requireNonNull(RepositorySizeAPI.all())
+                .stream()
+                .filter(r -> r.isApplicableTo(repoUrl))
+                .collect(Collectors.toList());
+
+        if (acceptedRepository.size() > 0) {
+            try {
+                for (RepositorySizeAPI repo: acceptedRepository) {
+                    long size = repo.getSizeOfRepository(repoUrl, context, credentialsId);
+                    if (size != 0) { sizeOfRepo = size; }
+                }
+            } catch (Exception e) {
+                LOGGER.log(Level.INFO, "Not using performance improvement from REST API: " + e.getMessage());
+                return false;
+            }
+            return sizeOfRepo != 0; // Check if the size of the repository is zero
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Recommend a git implementation on the basis of the given size of a repository
+     * @param sizeOfRepo: Size of a repository (in KiBs)
+     * @return a git implementation, "git" or "jgit"
+     */
+    private String determineSwitchOnSize(Long sizeOfRepo) {
+        if (sizeOfRepo != 0L) {
+            if (sizeOfRepo >= SIZE_TO_SWITCH) {
+                return "git";
```

Comment:
I think it would be better to define the names of Implementations as constants rather than hard coding them. Can be reused at other places too.

Reply:
Great point! Because of this implementation: https://github.com/jenkinsci/git-plugin/blob/e21bf121ee2ff4129676d097c88e2205b9b4ff6f/src/main/java/hudson/plugins/git/util/GitUtils.java#L59-L83

We should bare in mind:
- https://github.com/jenkinsci/git-client-plugin/blob/3a2230e6c3819d6110aebf888aa58d9cced77fab/src/main/java/org/jenkinsci/plugins/gitclient/JGitTool.java#L52-L55 is the correct constant for jgit
- What about JGit with Apache? https://github.com/jenkinsci/git-client-plugin/blob/3a2230e6c3819d6110aebf888aa58d9cced77fab/src/main/java/org/jenkinsci/plugins/gitclient/JGitApacheTool.java#L41-L44
- The constant "git" might be a problem. By default the installation name is "Default", but any other installation might have any name, such as "my-git" or "nothing-to-do-with-this-tool". Because of https://github.com/jenkinsci/git-plugin/blob/e21bf121ee2ff4129676d097c88e2205b9b4ff6f/src/main/java/hudson/plugins/git/util/GitUtils.java#L63-L69 then no "git" implementation will be found and the Default will be returned. If we change the Default name to "other-default", then we will get no installation. Do we care for those cases?